src/Microsoft.AspNetCore.Mvc.TagHelpers/CacheTagHelper.cs

```diff
@@ -174,29 +176,58 @@ internal MemoryCacheEntryOptions GetMemoryCacheEntryOptions()
         {
             var content = await output.GetChildContentAsync();
 
-            var stringBuilder = new StringBuilder();
-            using (var writer = new StringWriter(stringBuilder))
+            using (var writer = new CharBufferTextWriter())
             {
                 content.WriteTo(writer, HtmlEncoder);
+                return new CharBufferHtmlContent(writer.Buffer);
+            }
+        }
+
+        private class CharBufferTextWriter : TextWriter
+        {
+            public CharBufferTextWriter()
+            {
+                Buffer = new PagedCharBuffer(CharArrayBufferSource.Instance);
+            }
+
+            public override Encoding Encoding => Null.Encoding;
+
+            public PagedCharBuffer Buffer { get; }
+
+            public override void Write(char value)
+            {
+                Buffer.Append(value);
+            }
+
+            public override void Write(char[] buffer, int index, int count)
+            {
+                Buffer.Append(buffer, index, count);
             }
 
-            return new StringBuilderHtmlContent(stringBuilder);
+            public override void Write(string value)
+            {
+                Buffer.Append(value);
+            }
         }
 
-        private class StringBuilderHtmlContent : IHtmlContent
+        private class CharBufferHtmlContent : IHtmlContent
         {
-            private readonly StringBuilder _builder;
+            private readonly PagedCharBuffer _buffer;
 
-            public StringBuilderHtmlContent(StringBuilder builder)
+            public CharBufferHtmlContent(PagedCharBuffer buffer)
             {
-                _builder = builder;
+                _buffer = buffer;
             }
 
             public void WriteTo(TextWriter writer, HtmlEncoder encoder)
             {
-                for (var i = 0; i < _builder.Length; i++)
+                var length = _buffer.Length;
+
+                for (var i = 0; i < _buffer.Pages.Count; i++)
                 {
-                    writer.Write(_builder[i]);
+                    var pageLength = Math.Min(length, PagedCharBuffer.PageSize);
+                    writer.Write(_buffer.Pages[i], 0, pageLength);
+                    length -= pageLength;
```

Comment:
This ends up being problematic when `_buffer.Length` is a multiple of `PageSize`.  We could do mod PageSize + 1, but IMO that's not as intuitive.